### PR TITLE
sql: introduce large_full_scan_rows session variable

### PR DESCRIPTION
**Pull request by michae2**
This commit adds a new session variable (as well as the corresponding
cluster setting) `large_full_scan_rows` that determines the limit on
the table size for which `disallow_full_table_scans` considers the full
scans as "large" and rejects such queries. Note that when stats aren't
available, all scans are considered "large". This adds more granularity
to `disallow_full_table_scans` guardrail so that it would reject only
"large" scans. Note that the internally-issued queries are still
unaffected. The default value for the new variable is 1000 and setting
it to 0 will revert to the previous behavior where all full table/index
scans are rejected.

Release justification: low-risk improvement to the existing
functionality.

Release note (ops change): New session variable `large_full_scan_rows`
(as well as the corresponding cluster setting
`sql.defaults.large_full_scan_rows`) are introduced. This setting
determines which tables are considered "large" for the purposes of
enabling `disallow_full_table_scans` feature to reject full table/index
scans only of "large" table. The default value for the new setting is
1000, and in order to reject all full table/index scans (the previous
behavior) one can set the new setting to 0. Internally-issued queries
aren't affected, and the new setting has no impact when
`disallow_full_table_scans` feature is not enabled.

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
Looking good so far!

> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-:-MhzA9zm7jlQimh3t-1W:byq9y8d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/exec_util.go, line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-Mhz86jc61FO76b7__Ba:-Mhz86jc61FO76b7__Bb:b825hjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L588)):*
> ```Go
> var largeFullScanRows = settings.RegisterFloatSetting(
> 	"sql.defaults.large_full_scan_rows",
> 	"default value for large_full_scan_rows session setting",
> ```

can you add a bit about what the setting is for?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/exec_util.go, line 586 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MhzCRih2Kn4HPZEHns1:-MhzCRii-g8SDuLUeVJC:bpf41jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L586)):*
> ```Go
> ).WithPublic()
> 
> var largeFullScanRows = settings.RegisterFloatSetting(
> ```

I'm curious about the usage of "floats" rather than "ints" here.

Update: I see, it is because of the fact that we're using floats in the optimizer estimate. It'd be worth adding a quick comment about that here.

**Review by michae2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/exec_util.go, line 586 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MhzCRih2Kn4HPZEHns1:-MhzLAlO2JQA0hYH-Yl_:b-6wwhmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L586)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

That's a good point -- Is there a reason you don't want to use int64 and cast it to a float?
</blockquote></details>

Just my own paranoia that a day will come when we have two plans, each with full scans estimated as 0.4 rows and 0.5 rows respectively and will want to disable one but not the other. Maybe that's not possible with the current row estimates, it just seemed unlucky to tempt fate. I can change it if you want me to.

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/exec_util.go, line 586 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MhzCRih2Kn4HPZEHns1:-Mi6hDqvEjn62vlK6mWp:be2ancf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L586)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Just my own paranoia that a day will come when we have two plans, each with full scans estimated as 0.4 rows and 0.5 rows respectively and will want to disable one but not the other. Maybe that's not possible with the current row estimates, it just seemed unlucky to tempt fate. I can change it if you want me to.
</blockquote></details>

Ok, I'm fine with leaving it as a float, but a comment would be good as @yuzefovich suggested.

**Review by michae2 — COMMENTED** [review]
This is still just the first commit, but I wanted to rebase before going on vacation. No need to take a look yet.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-:-MiFOabl47Bpe7xD9hGl:baszwvv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/exec_util.go, line 586 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MhzCRih2Kn4HPZEHns1:-MiFO2xHESzN0iiCOmnL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L586)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Ok, I'm fine with leaving it as a float, but a comment would be good as @yuzefovich suggested.
</blockquote></details>

Done.
___
*[pkg/sql/exec_util.go, line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-Mhz86jc61FO76b7__Ba:-MiFOSVNE8kUMYqU4dgN:b-376vj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L588)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

can you add a bit about what the setting is for?
</blockquote></details>

Yes, I will as soon as I know exactly! :laughing:

**Review by yuzefovich — COMMENTED** [review]
I extended the commit to also take advantage of the newly introduced session variable. I think it now implements the desired modification of `disallow_full_table_scans` guardrail to reject only "large" scans. I adjusted the default value for the new variable to 1000, and I think we should be backporting it with the default value of 0 so that the previous behavior of rejecting all full table/index scans, regardless of the size, is preserved on 21.1 branch.

RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-:-MiTm8z8B99vHqV3G-J0:b-4n7kn0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/exec_util.go, line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-Mhz86jc61FO76b7__Ba:-MiTlwcI45NjLlzfGbbE:b33zx3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab9b00f26c5d7db3dce545ae6c429d74d30d9fb0/pkg/sql/exec_util.go#L588)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Yes, I will as soon as I know exactly! :laughing:
</blockquote></details>

I expanded the comment.

**Comment by yuzefovich** [comment]
Another update is that in the [functional spec](https://docs.google.com/document/d/1fWQ-c8YT_JJiupOze55xZ3nyrz2wGTtJ7TJKRqh0tnQ/edit#) we also wanted to do
```
1. Ignore the setting when performing a full table scan of internal tables (to allow SHOW queries to succeed).
```
However, I chatted with @vy-ton, and we agreed that the current behavior of prohibiting all large scans issued by the user, regardless of whether the scan is of an internal or a user table, is reasonable, so we won't make any changes to that logic. The `disallow_full_table_scans` guardrail does already allow all queries that are internally-issued by CRDB.

One of our customers mentioned that when they turned on the setting (`disallow_full_table_scans`), `SHOW` commands stopped working - I couldn't easily reproduce this issue (I played with `SHOW jobs` a bit), and the newly added `large_full_scan_rows` knob, when set to a reasonably large value, is likely to kick in and allow `SHOW` commands even if previously they were rejected. Thus, I think it's ok to not implement that part of the function spec. (On the implementation side it might be quite annoying to determine whether the large scan is of an internal or of a user table, so it could be somewhat risky to backport.)

**Comment by yuzefovich** [comment]
@rytaft friendly ping - it'd be nice to get this in today if possible.

One open question is whether we want to add a metric for the number of queries that are rejected by `disallow_full_table_scans` guardrails. We currently have a metric for the number of full table/index scans but nothing for the erring case of the guardrail, so it might be nice to add that, cc @vy-ton.

**Comment by vy-ton** [comment]
+1 to adding a new metric

**Comment by yuzefovich** [comment]
Added a metric.

**Review by rytaft — APPROVED** [review]
:lgtm:

Sorry for the delay!

> Reviewed 4 of 11 files at r2, 9 of 13 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-:-MicICuAED3YNEkRvYew:b-bhny20)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/exec_util.go, line 635 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MicGRa353loSdsHKOM8:-MicGRa45oiKKDCjlwOc:bnrqv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/70243c5371c39022d8c50979763cbda7d4f13322/pkg/sql/exec_util.go#L635)):*
> ```Go
> 	"sql.defaults.large_full_scan_rows",
> 	"default value for large_full_scan_rows session setting which determines "+
> 		"the limit on the table size full scans of which are allowed when "+
> ```

nit: I'd rephrase this a bit. Maybe "which determines the maximum table size allowed for a full scan when disallow_full_table_scans is set to true"

**Review by yuzefovich — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-:-MicKUuh-vqlY46Dly4v:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/exec_util.go, line 635 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69371#-MicGRa353loSdsHKOM8:-MicKGyN8vUJUDkP3cyu:b-86mr6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/70243c5371c39022d8c50979763cbda7d4f13322/pkg/sql/exec_util.go#L635)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: I'd rephrase this a bit. Maybe "which determines the maximum table size allowed for a full scan when disallow_full_table_scans is set to true"
</blockquote></details>

I like it, done.

**Review by rytaft — APPROVED** [review]
> Reviewed 8 of 8 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69371)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

**Comment by yuzefovich** [comment]
This should be backported after #70175 to reduce the merge conflicts.

**Comment by yuzefovich** [comment]
#70175 has just been merged.
